#! /bin/sh -e

# All lines beginning with `# DP:' are a description of the patch.
# DP: Description: Fix posix timer SIGEV_THREAD notification broken
# DP: Related bugs: #259878
# DP: Dpatch author: GOTO Masanori <gotom@debian.org>
# DP: Patch author: Jakub Jelinek <jakub@redhat.com>
# DP: Upstream status: In CVS
# DP: Status Details: 
# DP: Date: 2004-07-22

PATCHLEVEL=0

if [ $# -ne 2 ]; then
    echo >&2 "`basename $0`: script expects -patch|-unpatch as argument"
    exit 1
fi
case "$1" in
    -patch) patch -d "$2" -f --no-backup-if-mismatch -p$PATCHLEVEL < $0;;
    -unpatch) patch -d "$2" -f --no-backup-if-mismatch -R -p$PATCHLEVEL < $0;;
    *)
	echo >&2 "`basename $0`: script expects -patch|-unpatch as argument"
	exit 1
esac
exit 0

# append the patch here and adjust the -p? flag in the patch calls.
--- nptl/sysdeps/unix/sysv/linux/timer_routines.c	2003-04-21 16:13:40.000000000 +0900
+++ nptl/sysdeps/unix/sysv/linux/timer_routines.c	2004-07-22 00:54:39.000000000 +0900
@@ -1,4 +1,4 @@
-/* Copyright (C) 2003 Free Software Foundation, Inc.
+/* Copyright (C) 2003, 2004 Free Software Foundation, Inc.
    This file is part of the GNU C Library.
    Contributed by Ulrich Drepper <drepper@redhat.com>, 2003.
 
@@ -17,6 +17,7 @@
    write to the Free Software Foundation, Inc., 59 Temple Place - Suite 330,
    Boston, MA 02111-1307, USA.  */
 
+#include <errno.h>
 #include <setjmp.h>
 #include <signal.h>
 #include <stdbool.h>
@@ -61,7 +62,19 @@
     {
       siginfo_t si;
 
-      if (sigwaitinfo (&ss, &si) > 0)
+      /* sigwaitinfo cannot be used here, since it deletes
+        SIGCANCEL == SIGTIMER from the set.  */
+
+      int oldtype = LIBC_CANCEL_ASYNC ();
+
+      /* XXX The size argument hopefully will have to be changed to the
+        real size of the user-level sigset_t.  */
+      int result = INLINE_SYSCALL (rt_sigtimedwait, 4, &ss, &si, NULL,
+                                  _NSIG / 8);
+
+      LIBC_CANCEL_RESET (oldtype);
+
+      if (result > 0)
 	{
 	  if (si.si_code == SI_TIMER)
 	    {
